vyper/builtin_functions/functions.py

```diff
@@ -303,6 +319,22 @@ def fetch_call_return(self, node):
 
         return return_type
 
+    def infer_arg_types(self, node):
+        validate_call_args(node, 3)
+
+        slice_type = get_possible_types_from_node(node.args[0]).pop()
+        if not isinstance(slice_type, (BytesAbstractType, StringDefinition)):
```

Comment:
can we use `validate_expected_type` here?

Reply:
Since the type of the slice node has been derived, I thought that making another call to `validate_expected_type`, which then calls `get_possible_types_from_node` again, would be inefficient. I think one way is to add a kwarg to `validate_expected_type` which takes in an optional list of `given_types`. If this kwarg is supplied, then there is no need to run `get_possible_types_from_node`.

Reply:
so long as `get_possible_types_from_node` is idempotent, let's worry about efficiency later :). i don't think `get_possible_types_from_node` is a hotspot, and if it is, we can simply cache the value instead of having to write duplicated code.